projects/plugins/boost/app/rest-api/endpoints/Update_Cloud_CSS.php

```diff
@@ -22,7 +45,57 @@ public function request_methods() {
 	}
 
 	public function response( $request ) {
-		return ( new Cloud_CSS() )->update_cloud_css( $request->get_params() );
+		$state          = new Critical_CSS_State();
+		$storage        = new Critical_CSS_Storage();
+		$params         = $request->get_params();
+		$providers      = empty( $params['providers'] ) || ! is_array( $params['providers'] ) ? array() : $params['providers'];
+		$api_successful = array( 'success' => true );
+
+		// If success is false, the whole Cloud CSS generation process failed.
+		if ( empty( $params['success'] ) ) {
+			if ( empty( $params['message'] ) || ! is_string( $params['message'] ) ) {
+				$error = __( 'An unknown error occurred', 'jetpack-boost' );
+			} else {
+				$error = $params['message'];
+			}
+
+			$state->set_error( $error );
+			$state->save();
+
+			return $api_successful;
+		}
+
+		// Update each provider.
+		foreach ( $providers as $provider_key => $result ) {
+			if ( ! isset( $result['data'] ) ) {
+				return new \WP_Error( 'invalid_data', __( 'Invalid request; missing data element', 'jetpack-boost' ) );
+			}
+
+			$data = $result['data'];
+
+			// Success
+			if ( ! empty( $result['success'] ) && ! empty( $data['css'] ) && is_string( $data['css'] ) ) {
+				$storage->store_css( $provider_key, $data['css'] );
+				$state->set_provider_success( $provider_key );
+				continue;
+			}
+
+			// Collate URL errors.
+			$urls   = empty( $data['urls'] ) || ! is_array( $data['urls'] ) ? array() : $data['urls'];
+			$errors = array_map(
+				function ( $url ) use ( $urls ) {
+					return array_merge( array( 'url' => $url ), $urls[ $url ] );
+				},
+				array_keys( $data['urls'] )
+			);
```

Comment:
This is a bit confusing, maybe just use `foreach` instead? There's also URLs array above that's only partially used in the anonymous function.

Reply:
Sure; simplified :)